keps/sig-node/24-apparmor-ga/README.md

```diff
@@ -0,0 +1,682 @@
+Profiles have to be available on each node whereas the container runtime ensures
+that the profile is loaded when specified at pod or PSP level. Profiles can be
+specified per-container via the pod’s metadata annotation:
+
+```
+container.apparmor.security.beta.kubernetes.io/<container_name>: {unconfined,runtime/default,localhost/<profile>}
+```
+
+The feature has been more or less unchanged ever since. Also note that the
+addition predates feature gates or our modern concept of feature lifecycle. So,
+even though the annotations include `beta` in the key, this is entirely useable
+on any production GA cluster.
+
+The main motivation behind this KEP is to promote the AppArmor feature to GA.
+
+_NOTE: Seccomp was in a very similar state, but with some subtle differences.
+Promoting Seccomp to GA will be covered by a [separate
+KEP](/keps/sig-node/20190717-seccomp-ga.md)._
+
+### Goals
+
+- Declare AppArmor as GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace AppArmor annotations
+- Deprecate the AppArmor annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get AppArmor to GA, therefore all
+functional enhancements are out of scope, including:
+
+- Defining any standard "Kubernetes branded" AppArmor profiles
+- Formally specifying the AppArmor profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside of the node
+- Changing the semantics around AppArmor support
+- Windows support
+
+## Proposal
+
+AppArmor is not available on every Linux distribution. Beside this, container
+runtimes have AppArmor as compile-time feature which may be disabled as well.
+With the GA API we do not change the error handling and behave exactly the same
+as the current error propagation paths.
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with.
+There are no changes planned to the current `PodSecurityPolicy` API as part of
+this KEP.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable, except in `PodTemplates`.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container. If AppArmor options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
```

Comment:
is this stricter than the current annotation validation (I think it is)? if so, that means currently working annotations could not be copied into the field

Reply:
Is it about the profile length? That's the AppArmor limitation rather than any validation we do in Kubernetes. So I guess it's not stricter.

Reply:
The question is what this API-level validation would do with existing deployment/replicasets/pods which have an annotation longer than 253 characters.

Current behavior:
- API operations are not blocked
- On nodes that don't support apparmor, those would run fine
- On nodes that do support apparmor, those would fail with a runtime error about the path being too long? (verify)

Proposed new behavior:
- translation between annotation and field in pod creation would fail on field validation?

Reply:
If a profile name has more than 253 characters, then it would not possible to load it at all in memory. This means a runtime will fail to look it up. I think we should not enforce this by the API…should I remove the comment?

Reply:
> This means a runtime will fail to look it up. I think we should not enforce this by the API…should I remove the comment?

It can be helpful to describe, but should not be confused with an API-enforced field limit

Reply:
Alright, I removed it to avoid confusion. :+1: